readme.md
# Audiomoth Visualizer

Node.js pipeline in the console for spectrograms creation and [Birdnet](https://github.com/kahst/BirdNET) detection for your [Audiomoth](https://www.openacousticdevices.info/audiomoth) sessions.  
It will create a movie per audiofile with a spectrogram as image and utilizes birdnet AI to detect bird species.  
The HTML page lists all movies and renders the birdnet detections.

Example result: [https://audiomoth.ivo-bathke.name](https://audiomoth.ivo-bathke.name/)

![](audiomoth-visualizer.png)

![](asciinema.svg)

## Installation

You will need `docker`,`sox` and `ffmpeg` on your machine.  
On Linux install it like this:

    sudo apt install sox ffmpeg

For docker installation see here: https://docs.docker.com/engine/install/.  
Birdnet detection software will run inside a docker container.

## Usage

#### Globally via npm

Install:

    npm install --global audiomoth-visualizer

Run:

    audiomoth-visualizer <path-to/my-audiomoth-session> <destination-dir> --title="Session title" --locale=de-DE --timezone=Europe/Berlin --lat=50.8754444 --lng=6.8380488

#### Running on-demand:

Run:

    npx audiomoth-visualizer <path-to/my-audiomoth-session> <destination-dir> --title="Session title" --locale=de-DE --timezone=Europe/Berlin --lat=50.8754444 --lng=6.8380488

Using npx you can run the script without installing it first.

#### Local dev environment

Clone this repository:

    git clone git@github.com:ivoba/audiomoth-visualizer.git

Install dependencies:

    npm i

    node index.mjs <path-to/my-audiomoth-session> <destination-dir> --title="Session title" --locale=de-DE --timezone=Europe/Berlin --lat=50.8754444 --lng=6.8380488

Then open the generated `files/my-audiomoth-session/index.html` in your favorite Browser by doubleclick or via console:

    chromium <path-to/my-audiomoth-session>/index.html

## Development

There are entrypoints to test birdnet AI or HTML generation separately:

HTML generation:

    node_modules/.bin/zx html.mjs <path-to/my-audiomoth-session>

Birdnet detection:

    node_modules/.bin/zx birdnet.mjs <path-to/my-audiomoth-session>

## Release

    npm run release

## Credits

Heavily inspired by:  
https://github.com/nwolek/audiomoth-scripts

## Todo

- responsive images
- responsive video
- with https://www.tensorflow.org/js/
- docker image in npx? will the image be pulled?
- docker user permissions on birdnet files
  ` --user $(id -u):$(id -g)`
  user should be used in docker image
  https://github.com/kahst/BirdNET/issues/23
- tests: https://dev.to/shadowtime2000/testing-command-line-tools-47bk
- https://engineering.britebill.com/2018/10/30/end-to-end-testing-cli-apps-jest.html
- print percentage of detection
- create a clean task to delete all files with no detection
- scroll to top link on bottom
- paging?
- fix Changelog generation
- input date -> output date options


CHANGELOG.md
## [1.4.1](https://github.com/ivoba/audiomoth-visualizer/compare/1.3.2...1.3.3) (2022-05-25)

- removed release-it, cosmetics

## [1.4.0](https://github.com/ivoba/audiomoth-visualizer/compare/1.3.2...1.3.3) (2022-05-25)

- switched to tflite model of https://github.com/kahst/BirdNET-Analyzer in docker image ivoba/birdnet

## [1.3.3](https://github.com/ivoba/audiomoth-visualizer/compare/1.3.2...1.3.3) (2022-05-25)

## [1.3.2](https://github.com/ivoba/audiomoth-visualizer/compare/1.3.1...1.3.2) (2022-05-25)

### Bug Fixes

- added chunking ([f4fce30](https://github.com/ivoba/audiomoth-visualizer/commit/f4fce309aa18b85dced944cf18c16e2995f59c18))

## [1.3.1](https://github.com/ivoba/audiomoth-visualizer/compare/1.3.0...1.3.1) (2022-05-25)

### Bug Fixes

- added @semantic-release/release-notes-generator ([ab46298](https://github.com/ivoba/audiomoth-visualizer/commit/ab46298072e78b81f4fd16b6499a6d96b89d808e))

# [1.3.0](https://github.com/ivoba/audiomoth-visualizer/compare/1.2.0...1.3.0) (2022-05-25)

### Features

- improved HTML, fixed image generation, deps update ([2d3ba85](https://github.com/ivoba/audiomoth-visualizer/commit/2d3ba8598aa96d1ac8d7a8ab471f9da1a0e49651))

# [1.2.0](https://github.com/ivoba/audiomoth-visualizer/compare/1.1.0...1.2.0) (2022-04-28)

### Features

- **views:** added esc key to close window, deps update & release-it fixes ([cb0d12d](https://github.com/ivoba/audiomoth-visualizer/commit/cb0d12d3ed2c5871176489c509c24c59ca615fa8))

# [1.1.0](https://github.com/ivoba/audiomoth-visualizer/compare/1.0.0...1.1.0) (2022-03-07)

### Features

- **Links:** added anchor links per track ([7dd9d1c](https://github.com/ivoba/audiomoth-visualizer/commit/7dd9d1cb147739e71378edf557af2dbeef76a3d0))

# [1.0.0](https://github.com/ivoba/audiomoth-visualizer/compare/0.0.5...1.0.0) (2022-02-16)

### Features

- added --week to birdnet ([fc08e92](https://github.com/ivoba/audiomoth-visualizer/commit/fc08e92e75324eb5eead2f238cc335182d22285d))
- **Maps:** added staticmaps & lat/lon to birdnet ([b920877](https://github.com/ivoba/audiomoth-visualizer/commit/b920877848d15564ef8a8dbb7ac0ad76c306d6d8))

## [0.0.5](https://github.com/ivoba/audiomoth-visualizer/compare/0.0.4...0.0.5) (2022-02-10)

### Bug Fixes

- **readme:** added asciinema ([061abad](https://github.com/ivoba/audiomoth-visualizer/commit/061abadc0df809e357889336172b4ffa1de74a46))

## [0.0.4](https://github.com/ivoba/audiomoth-visualizer/compare/0.0.3...0.0.4) (2022-02-09)

### Bug Fixes

- **html:** fixed style.scss path ([42d0ef1](https://github.com/ivoba/audiomoth-visualizer/commit/42d0ef183586d91656c60229d2d54ad7cf13051a))

## [0.0.3](https://github.com/ivoba/audiomoth-visualizer/compare/0.0.2...0.0.3) (2022-02-09)

### Bug Fixes

- **cli:** dont rely on zx as entry ([0a57377](https://github.com/ivoba/audiomoth-visualizer/commit/0a573774fe48393cab41852fcb75bc883d40078c))

## [0.0.2](https://github.com/ivoba/audiomoth-visualizer/compare/0.0.1...0.0.2) (2022-02-09)

### Bug Fixes

- node engine constraint ([f528be1](https://github.com/ivoba/audiomoth-visualizer/commit/f528be1647571163b4e54488ad6992012149fc63))

## 0.0.1 (2022-02-09)


lib/html.mjs
import fs from "fs";
import posthtml from "posthtml";
import content from "posthtml-content";
import expressions from "posthtml-expressions";
import sass from "sass";
import { updateManifestWithBirdnetData } from "./birdnetManifest.mjs";

async function generateHTML(dir) {
  if (fs.existsSync(dir)) {
    const manifest = JSON.parse(
      fs.readFileSync(`${dir}/manifest.json`).toString()
    );
    await updateManifestWithBirdnetData(dir, manifest);

    const css = await sass.compile(
      new URL("../views/style.scss", import.meta.url).pathname
    );
    fs.writeFileSync(`${dir}/style.css`, css.css.toString());

    const dateFormatter = (str, locale, timeZone) => {
      const date = new Date(str);
      const df = new Intl.DateTimeFormat(locale, {
        day: "numeric",
        month: "numeric",
        year: "numeric",
        hour: "numeric",
        minute: "numeric",
        second: "numeric",
        timeZone: timeZone,
      });
      return df.format(date);
    };

    const contentPlugin = content({
      dateformat: (str) => dateFormatter(str),
      firstDate: (filesJson) => {
        const files = JSON.parse(filesJson);
        return dateFormatter(
          files[0].record_date,
          manifest.locale,
          manifest.time_zone
        );
      },
      lastDate: (filesJson) => {
        const files = JSON.parse(filesJson);
        return dateFormatter(
          files[files.length - 1].record_date,
          manifest.locale,
          manifest.time_zone
        );
      },
      filesCount: (filesJson) => {
        const files = JSON.parse(filesJson);
        return files.length;
      },
      roundConfidence: (confidence) => {
        return Math.round(confidence * 100) / 100;
      },
    });
    posthtml([expressions({ locals: manifest }), contentPlugin])
      .process(
        fs.readFileSync(new URL("../views/index.html", import.meta.url), "utf8")
      )
      .then((result) => fs.writeFileSync(`${dir}/index.html`, result.html));
  }
}

export { generateHTML };


lib/movieFrames.mjs
const createMovieFrames = async (
  dest,
  file,
  baseFileName,
  totalDuration,
  thumbMaxDuration,
  title,
  date,
  thumb
) => {
  let imageCount = 0;
  for (let i = 0; i < totalDuration; i = i + thumbMaxDuration) {
    let frameImage = `${dest}/tmp_${baseFileName}${imageCount}.png`;
    console.log(
      `making spectrogram from ${i} to ${i + thumbMaxDuration} seconds...`
    );
    // generate the initial .png spectrogram output from sox
    let annotation = `${title}\n ${date.mtime.toISOString()}`;
    await $`sox ${file} -n rate 24k trim 0 10 spectrogram -x 1136 -y 642 -z 96 -w hann -o -`
      //extend image with black background & add text to bottom of image
      .pipe(
        $`convert PNG:- -background black -gravity north -extent 1280x820 - | convert PNG:- -gravity south -fill white -pointsize 36 -annotate +0+10 ${annotation} ${frameImage}`
      );
    imageCount++;
  }
  const firstFrame = `${dest}/tmp_${baseFileName}0.png`;
  await $`convert ${firstFrame} -resize 300x170 ${thumb}`;
};

export { createMovieFrames };


lib/birdnetManifest.mjs
import fs from "fs";
import csvreader from "csv-reader";
import * as path from "path";

const updateManifestWithBirdnetData = async (dir, manifest) => {
  return await Promise.all(
    manifest.files.map(async (fileData) => {
      const birdnetFile = new URL(
        `${dir}/birdnet/${fileData.base}.BirdNET.selection.table.txt`,
        import.meta.url
      );
      if (!fs.existsSync(birdnetFile)) {
        return fileData;
      }
      return mergeFileData(fileData, birdnetFile);
    })
  );
};

const mergeFileData = async (fileData, birdnetFile) => {
  let inputStream = fs.createReadStream(birdnetFile, "utf8");
  let data = {};
  inputStream
    .pipe(
      new csvreader({
        delimiter: "\t",
        parseNumbers: true,
        parseBooleans: true,
        trim: true,
        skipHeader: true,
      })
    )
    .on("data", function (row) {
      let seconds = `${row[3]}-${row[4]}`;
      let timeData = {
        start: row[3],
        end: row[4],
        specie: row[8],
        rank: row[0],
        confidence: row[9],
      };
      if (data[seconds] === undefined) {
        data[seconds] = [timeData];
      } else {
        data[seconds].push(timeData);
      }
    })
    .on("end", function () {
      data = Object.keys(data)
        .sort()
        .reduce((obj, key) => {
          obj[key] = data[key];
          return obj;
        }, {});
      fileData.birdnet = data;
    });

  return new Promise(function (resolve, reject) {
    inputStream.on("end", () => resolve(fileData));
    inputStream.on("error", reject); // or something like that. might need to close `hash`
  });
};

const loadBirdnetFile = async (file, manifest) => {
  const basename = path.basename(file, "WAV");
  const dir = path.dirname(file);
  const birdnetFile = `${dir}/birdnet/${basename}BirdNET.selections.txt`;
  if (!fs.existsSync(birdnetFile)) {
    return null;
  }

  let inputStream = fs.createReadStream(birdnetFile, "utf8");
  let data = {};

  inputStream
    .pipe(
      new csvreader({
        delimiter: "\t",
        parseNumbers: true,
        parseBooleans: true,
        trim: true,
        skipHeader: true,
      })
    )
    .on("data", function (row) {
      let seconds = `${row[3]}-${row[4]}`;
      let timeData = {
        start: row[3],
        end: row[4],
        specie: row[8],
        rank: row[0],
        confidence: row[9],
      };
      if (data[seconds] === undefined) {
        data[seconds] = [timeData];
      } else {
        data[seconds].push(timeData);
      }
    })
    .on("end", function () {
      let obj = manifest.files.find((el) => el.base === basename.slice(0, -1));
      if (obj) {
        data.sort();
        obj.birdnet = data;
      }
    });

  return new Promise(function (resolve, reject) {
    inputStream.on("end", () => resolve(manifest));
    inputStream.on("error", reject); // or something like that. might need to close `hash`
  });
};

export { updateManifestWithBirdnetData, loadBirdnetFile };


index.mjs
#!/usr/bin/env node

import * as path from "path";
import "zx/globals";
import { createMovieFrames } from "./lib/movieFrames.mjs";
import { birdnet } from "./lib/birdnet.mjs";
import { updateManifestWithBirdnetData } from "./lib/birdnetManifest.mjs";
import { generateHTML } from "./lib/html.mjs";
import StaticMaps from "staticmaps";

const dir = argv._[0];
const dest = path.resolve(argv._[1] ?? dir);

if (!fs.existsSync(dir)) {
  console.log("no such dir ", dir);
  process.exit(1);
}
const title = argv.title ?? "";
const locale = argv.locale ?? "de-DE";
const timeZone = argv.timezone ?? "Europe/Berlin";
const lat = argv.lat ?? null;
const lng = argv.lng ?? null;

const allFiles = fs.readdirSync(dir);
let files = allFiles.filter(function (elm) {
  return elm.match(/.*\.(WAV)/gi);
});
if (files.length == 0) {
  console.log("no files found", dir);
  process.exit(1);
}

const audiosDir = "audios";
const imagesDir = "images";
const thumbsDir = "thumbs";
const moviesDir = "movies";
const audiosPath = `${dest}/${audiosDir}`;
const imagesPath = `${dest}/${imagesDir}`;
const thumbsPath = `${dest}/${imagesDir}/${thumbsDir}`;
const moviesPath = `${dest}/${moviesDir}`;

if (!fs.existsSync(dest)) {
  fs.mkdirSync(dest);
}
if (!fs.existsSync(audiosPath)) {
  fs.mkdirSync(audiosPath);
}
if (!fs.existsSync(imagesPath)) {
  fs.mkdirSync(imagesPath);
}
if (!fs.existsSync(thumbsPath)) {
  fs.mkdirSync(thumbsPath);
}
if (!fs.existsSync(moviesPath)) {
  fs.mkdirSync(moviesPath);
}

let map;
if (lat && lng) {
  map = new StaticMaps({
    width: 400,
    height: 200,
  });
  const marker = {
    img: new URL(`./views/marker.png`, import.meta.url).pathname,
    offsetX: 12,
    offsetY: 48,
    width: 24,
    height: 24,
    coord: [lng, lat],
  };
  map.addMarker(marker);
  await map.render([lng, lat], 13);
  await map.image.save(`${dest}/map.png`, { compressionLevel: 9 });
}

const manifest = {
  title: title,
  location: {
    lat: lat,
    lng: lng,
    map_image: map,
  },
  locale: locale,
  time_zone: timeZone,
  files: [],
};

const chunk = (arr, size) =>
  arr.reduce(
    (acc, e, i) => (
      i % size ? acc[acc.length - 1].push(e) : acc.push([e]), acc
    ),
    []
  );

const chunks = chunk(files, 10);

for (const chunk of chunks) {
  await Promise.all(
    chunk.map(async (audioFile) => {
      let file = `${dir}/${audioFile}`;
      let destAudioFile = `${audiosDir}/${audioFile}`;
      fs.copyFileSync(file, `${dest}/${destAudioFile}`);
      let baseFileName = path.basename(`${dir}/${file}`, ".WAV");
      let thumbFile = `${baseFileName}.png`;
      let thumb = `${thumbsPath}/${thumbFile}`;
      let movieFile = `${baseFileName}.mp4`;
      let movie = `${moviesPath}/${movieFile}`;
      let totalDuration = parseInt(await $`soxi -D ${file}`);
      let thumbMaxDuration = 30;
      let date = await fs.stat(file);
      if (totalDuration < thumbMaxDuration) thumbMaxDuration = totalDuration;
      await createMovieFrames(
        dir,
        file,
        baseFileName,
        totalDuration,
        thumbMaxDuration,
        title,
        date,
        thumb
      );
      console.log(`creating full length movie: ${movie}`);
      await $`ffmpeg -loglevel panic -framerate 1/${thumbMaxDuration} -i ${dir}/tmp_${baseFileName}%d.png -i ${file} -c:v libx264 -y -crf 23 -pix_fmt yuv420p ${movie}`;
      manifest.files.push({
        base: baseFileName,
        audio: `${destAudioFile}`,
        thumb: `${imagesDir}/${thumbsDir}/${thumbFile}`,
        movie: `${moviesDir}/${movieFile}`,
        record_date: date.mtime.toISOString(),
      });
    })
  );
  console.log("Processed chunk:", chunk);
}
console.log("Removing tmp files...");
let regex = new RegExp("^tmp_.*png$");
fs.readdirSync(dir)
  .filter((f) => regex.test(f))
  .map((f) => fs.unlinkSync(`${dir}/${f}`));

// sort files in manifest
manifest.files.sort((a, b) =>
  a.audio > b.audio ? 1 : b.audio > a.audio ? -1 : 0
);

await birdnet(
  dest,
  audiosDir,
  "birdnet",
  manifest,
  locale.substr(0, 2).toLowerCase()
);

await updateManifestWithBirdnetData(dest, manifest);
await fs.writeJson(`${dest}/manifest.json`, manifest);

await generateHTML(dest);

console.log("Done: file://" + dest + "/index.html");
